### ISSUE TITLE
ama-metrics-operator-targets consuming more and more cluster memory

### ISSUE THREAD

**Issue by ChrisJD-VMC** [Azure/AKS#4509]
**Describe the bug**
I don't know if this is the correct place or this, if it's not please advise where to direct this issue.
tldr; ama-metrics-operator-targets seems to have a memory leak (I assume it's not designed to slowly consume more and more RAM).

I got alerts from both AKS clusters I run this morning that a container in each had been OOM killed. Some investigation revealed that the containers in question were both the ama-metrics-operator-targets (Azure Managed Prometheus monitoring related, to my understanding).

Looking at the memory usage for those containers in Prometheus I can see a ramp up in memory usage over the course of probably a bit more than a week followed by the containers being killed at about 2GB of ram usage. The memory use then drops back to 60-70MB and then starts climbing again.

This is the first time this has happened. We've been using Azure Managed Prometheus for about 3 months. Given the rate the RAM usage is increasing at I assume some kind of new issue is causing this. Probably introduced in the last couple of weeks. We have not made any changes to either clusters configuration for several months. And one of the clusters hasn't had any container changes deployed by us for 3 months. Both are configured to auto update for minor cluster versions.

**To Reproduce**
I assume just having a cluster configured with Prometheus monitoring is enough. 

**Expected behavior**
ama-metrics-operator-targets container RAM usage does not continuously grow over time.

**Screenshots**
7 days ago
![image](https://github.com/user-attachments/assets/dca7eb30-148a-47b9-9c61-39cc3e233dcb)
Last night
![image](https://github.com/user-attachments/assets/05efe5fd-c0f5-4dd2-af00-974ca506d026)
After the OOM kill occurred
![image](https://github.com/user-attachments/assets/a7437196-34f9-4f48-b55a-61f6661bea2a)
Climbing again
![image](https://github.com/user-attachments/assets/0451540f-31ee-43d5-8b06-f1d2790098ff)

**Environment (please complete the following information):**
CLI Version - 2.62.0
Kubernetes version - 1.29.7 and 1.30.3
Browser - chrome

**Additional Info**
Clusters are in two different regions. Connected using AMPLS to the same Azure Monitor Workspace. One Azure Managed Prometheus instance connected to the workspace. Data still appears to be being collected and can be viewed fine in Prometheus.

**Comment by boyko11** [comment]
Started at 24 Megs for the config-reader 4 days ago, today at 600 Megs:
![image](https://github.com/user-attachments/assets/6b3711ac-0ea8-4c18-ad6a-7de62010b830)
![image](https://github.com/user-attachments/assets/09499053-cb26-4899-8237-e4e0f6aab2d9)

**Comment by akari-m** [comment]
We have the similar issue. The memory usage keeps growing and drops down suddenly

**Comment by vishiy** [comment]
Hi - this is a known issue that we will be rolling out a fix for.

**Comment by JoeyC-Dev** [comment]
Hi @vishiy 
Do we have any ETA for the fix? Thanks.

**Comment by akari-m** [comment]
Hi @vishiy 

Is there any way we could free the memory of the ama-metrics-operator-targets pod like manually kill the pod?

Will killing the pod make any impacts in AKS cluster?

**Comment by boyko11** [comment]
> Hi @vishiy
> 
> Is there any way we could free the memory of the ama-metrics-operator-targets pod like manually kill the pod?
> 
> Will killing the pod make any impacts in AKS cluster?

@akari-m , I opened a support ticket for this issue. The support engineer had me delete all the pods whose name start with ama-*. The pods recreated automatically, we reclaimed over 2Gigs of memory, no impact to our application pods…

**Comment by deyanp** [comment]
Not only the memory consumption is an issue, but also the CPU consumption. Why is this pod taking 1732Mi in avg???

**Comment by iakovmarkov** [comment]
Happened today to our production cluster as well. CC @vishiy.

In general it seems like the Azure Monitor for AKS is not in a good shape. It's very convinient one-click deployment, but the stability and quality of the setup is pretty damn low for a paid product.

**Comment by ppanyukov** [comment]
This started happening to us on 16th of August 2024 (see chart), both in WestEurope and WestUS clusters, all by itself.  We are on 1.28.9 AKS version. 

The leak causes node reboots for us, and considering this component is runnning on the system nodepool, leads all sorts of bad effects.

@vishiy is there any ETA for the fix you mentioned, or is there anything else we can do to resolve the issue? 


![image](https://github.com/user-attachments/assets/cb496fbd-253a-42a2-a59e-29a6a6bf93c9)

**Comment by adejongh** [comment]
We have the same issue.

**Comment by AkariH** [comment]
Hello @vishiy . We have the same issue. Please assist

**Comment by twanbeeren** [comment]
This issue also causes OOMkills for us.

Temporary solution was to create a cronjob that restarts the operator every day so that it doesn't consume a lot of memory and causes issues on the nodes.

Please notify in here once it is resolved

**Comment by deyanp** [comment]
Can someone post a 1-liner kubectl command to create this cron job as a temporary workaround? ;)

**Comment by twanbeeren** [comment]
Cronjob

```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: kill-ama-operator-cj
  namespace: kube-system
spec:
  schedule: "0 6 * * *" # Runs every day at 6:00 AM
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: kill-pod
            image: bitnami/kubectl:latest
            command:
            - /bin/sh
            - -c
            - |
              POD=$(kubectl get pods -n kube-system -l rsName=ama-metrics-operator-targets -o jsonpath='{.items[0].metadata.name}')
              kubectl delete pod $POD -n kube-system
          restartPolicy: OnFailure
```

If you want to check the Cronjob deployment:
`kubectl create job test-job --from=cronjob/kill-ama-operator-cj -n kube-system`

If your job fails with a kubectl error you probably need to add a serviceaccount

**Comment by antiphon0** [comment]
Can confirm this is happening on our clusters as well.

**Comment by shiroshiro14** [comment]
> Hi - this is a known issue that we will be rolling out a fix for.

Is this ever going to be fixed or it would be my fever dream forever?

**Comment by deyanp** [comment]
@twanbeeren , I allowed myself to extend your version: 
 
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: kill-ama-metrics-operator-targets-cj-sa
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: kube-system
  name: kill-ama-metrics-operator-targets-role
rules:
- apiGroups: [""]
  resources: ["pods"]
  verbs: ["get", "list", "delete"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: kill-ama-metrics-operator-targets-cj-sa-binding
  namespace: kube-system
subjects:
  - kind: ServiceAccount
    name: kill-ama-metrics-operator-targets-cj-sa
    namespace: kube-system
roleRef:
  kind: Role
  name: kill-ama-metrics-operator-targets-role
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: batch/v1
kind: CronJob
metadata:
  name: kill-ama-metrics-operator-targets-cj
  namespace: kube-system
spec:
  schedule: "0 6 * * *" # Runs every day at 6:00 AM
  jobTemplate:
    spec:
      template:
        spec:
          serviceAccountName: kill-ama-metrics-operator-targets-cj-sa
          containers:
          - name: kill-pod
            image: bitnami/kubectl:latest
            command:
            - /bin/sh
            - -c
            - |
              POD=$(kubectl get pods -n kube-system -l rsName=ama-metrics-operator-targets -o jsonpath='{.items[0].metadata.name}')
              kubectl delete pod $POD -n kube-system
          restartPolicy: OnFailure
  successfulJobsHistoryLimit: 1
  failedJobsHistoryLimit: 1

```

**Comment by rashmichandrashekar** [comment]
The fix for this is rolling out currently. It should roll out to all regions by 09/30 in all regions.

**Comment by rashmichandrashekar** [comment]
> > The fix for this is rolling out currently. It should roll out to all regions by 09/30.
> > I am still facing this issue.

@sivashankaran22 - Could you provide me your cluster id?

**Comment by adejongh** [comment]
> The fix for this is rolling out currently. It should roll out to all regions by 09/30.

What is the version of the fix that your are rolling out, so we can check that has been deployed?

We still see very high memory use for all the "ama-" pods, which is really not acceptable.

<img width="667" alt="image" src="https://github.com/user-attachments/assets/5a1f370d-3857-48cd-97f4-dea215532fbc">

**Comment by KodandaRamu1** [comment]
> The fix for this is rolling out currently. It should roll out to all regions by 09/30.

Still facing the same issue, for the below cluster
 ResourceUri: /subscriptions/632b09d2-0e9a-463b-a5e9-2209635f3409/resourceGroups/pvad-prod-weu-rg/providers/Microsoft.ContainerService/managedClusters/pvad-prod-weu-aks

**Comment by ChrisJD-VMC** [comment]
Memory usage has been back to normal for a while for us. So I'm going to close this as it is resolved as far as I am concerned
